### PR TITLE
Add `prev_blockhash` validation to `CheckPoint`

### LINKED ISSUE #2021: Ensure `CheckPoint` chain methods validate and link via previous blockhash
**Describe the enhancement**  

#1582 made `CheckPoint`s generic so that we can attach `Header`s and `Block`s to a given `CheckPoint`. Since these newly supported data types have prev-blockhashes, this introduced a new invariant: prev-blockhash of the current checkpoint must match the (current_checkpoint_height-1)'s checkpoint's blockhash (if it exists).

Additionally, when merging chains with prev-blockhash data means even if two chain do not have a point-of-agreement, they may be able to connect if a chain's prev-blockhash matches a checkpoint in the other chain.

**Proposed Changes:**

* Extend `ToBlockHash` to have knowledge of whether the data contains a prev_blockhash.
  ```rust
  pub trait ToBlockHash {
      /* Other method */
    
      fn prev_blockhash(&self) -> Option<BlockHash>;
  }
  ```

* Modify `CheckPoint` to support having empty data. When a checkpoint with a prev-blockhash is inserted, also insert an empty-data checkpoint at height-1.

**Use case**  

Ensures consistent representation of data.

**Impact**

We catch problems earlier. Either as the chain-source when they construct the `CheckPoint` update, or when trying to apply the update to the receiving structures afterwards.

**Are you using BDK in a production project?**
- [x] Yes
- [ ] No
- [ ] Not yet, but planning to

**Which backend(s) are relevant (if any)?**

Yes, this will affect all chain-source implementations that construct checkpoint updates. This will also affect applying updates as the problem might be caught there.

**Project or organization (optional)**

I get funded by the BDKF and I also work for Frostsnap.

**Additional context**

Original comment: https://github.com/bitcoindevkit/bdk/pull/1582#issuecomment-3214838670

### PR DESCRIPTION
**Pull request by evanlinjin**
Closes #2021
Related to #2076
Replaces #2024
Replaces #2091

### Description

This PR adds `prev_blockhash` awareness to `CheckPoint`, enabling proper chain validation when merging checkpoint chains that store block headers or similar data with previous block hash information.

### Notes to the reviewers

This PR replaces some prior attempts:

* #2024 - where we made the `CheckPoint::data` optional - however this resulted in internal complexity and an API with annoying edge cases. The tests from this PR were still useful.

* #2091 - This second attempt had some good ideas, but was distracted from the goal of #2021. I mostly reused the `CheckPoint::insert` implementation of that PR.

### Changelog notice

```md
Added:
- `ToBlockHash::prev_blockhash()` - optional method to expose previous block hash
- `CheckPointEntry` - new type for iterating with `prev_blockhash` awareness, yielding "placeholder" entries for heights inferred from `prev_blockhash`
- `ApplyBlockError` - this is a new error type with two variants; `MissingGenesis` and `PrevBlockhashMismatch`. The second variant is a new error case introduced by `prev_blockhash` awareness.

Changed:
- `CheckPoint::push` - now errors when `prev_blockhash` conflicts with current tip (contiguous heights)
- `CheckPoint::insert` - now evicts/displaces checkpoints on `prev_blockhash` conflict
- `merge_chains` - now validates `prev_blockhash` consistency when merging
- `LocalChain<D>` generic parameter - relaxed constraint to `D: Clone` instead of `D: Copy`.

Fixed:
- `merge_chains` no longer replaces the genesis block.
```

#### All Submissions:

* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

### CODE REVIEW

File: crates/chain/src/local_chain.rs [hunk 1]
```diff
@@ -671,7 +676,8 @@ where
                     prev_orig_was_invalidated = false;
                     // OPTIMIZATION 2 -- if we have the same underlying pointer at this point, we
                     // can guarantee that no older blocks are introduced.
-                    if o.eq_ptr(u) {
+                    // TODO: Ensure this is correct!
```
Comment (evanlinjin) [inline crates/chain/src/local_chain.rs]: This is correct! We can remove this comment!

File: crates/chain/src/local_chain.rs
Comment (evanlinjin) [inline crates/chain/src/local_chain.rs:1]: As highlighted in f7eedf3f1e84e62d7480cb24de6f4a41204b9c76, we should do something about `apply_changeset_to_checkpoint`.

**Comment by evanlinjin** [comment]
This PR is actually ready for review now!

File: crates/chain/src/local_chain.rs [hunk 2]
```diff
@@ -485,6 +494,36 @@ impl<D> FromIterator<(u32, D)> for ChangeSet<D> {
     }
 }
 
+/// Error when applying blocks to a local chain.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ApplyBlockError {
+    /// Genesis block is missing or would be altered.
+    MissingGenesis,
```
Comment (nymius) [inline crates/chain/src/local_chain.rs:501]: I would split this error in two to keep a single semantic for each one: `MissingGenesis` and `ConflictingGenesis` or similar

File: crates/chain/src/local_chain.rs [hunk 3]
```diff
@@ -634,13 +697,18 @@ where
         match (curr_orig.as_ref(), curr_update.as_ref()) {
             // Update block that doesn't exist in the original chain
             (o, Some(u)) if Some(u.height()) > o.map(|o| o.height()) => {
-                changeset.blocks.insert(u.height(), Some(u.data()));
+                // Only append to `ChangeSet` when this is an actual checkpoint.
```
Comment (nymius) [inline crates/chain/src/local_chain.rs]: What's the definition of "actual checkpoint": one that isn't a Placeholder or one that contains some data?
Why would it be a checkpoint with no data but at the same time is not a Placeholder?
Wouldn't be better to rewrite this comment to not leave to the reader the definition of "actual checkpoint"?

File: crates/core/src/checkpoint.rs [hunk 4]
```diff
@@ -281,74 +310,119 @@ where
...
+            } else if cp.height() < height {
+                break Some(cp);
             }
 
-            if cp.height() < height {
-                break cp;
+            // Continue traversing down (if possible).
+            match cp.prev() {
+                Some(prev) => cp = prev,
+                None => break None,
```
Comment (nymius) [inline crates/core/src/checkpoint.rs:376]: When we hit `None` on this iterator, it means we have exhausted `CheckPoint` so we are beyond genesis block?

File: crates/core/tests/test_checkpoint.rs [hunk 5]
```diff
@@ -186,3 +186,387 @@ fn test_mtp_sparse_chain() {
...
+    // Verify height 6 is purged
+    assert!(
+        result.get(6).is_none(),
+        "checkpoint at 6 should be purged due to prev_blockhash conflict"
+    );
+
+    // Verify chain structure
+    assert_eq!(result.height(), 5, "tip should be at height 5");
+    // Should have: checkpoint 5 only
+    assert_eq!(
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs]: What's the difference between displaced and purged if 4 is not part of the chain?

File: crates/core/tests/test_checkpoint.rs [hunk 6]
```diff
@@ -186,3 +186,387 @@ fn test_mtp_sparse_chain() {
...
+        result.is_ok(),
+        "push should succeed for non-contiguous height"
+    );
+    let new_cp = result.unwrap();
+
+    // Verify the tip is at 105
+    assert_eq!(new_cp.height(), 105);
+    assert_eq!(new_cp.hash(), hash!("block_105"));
+
+    // Verify chain structure: 100, 105
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs]: This is not precisely verifying chain structure. An explicit check for both `CheckPoint`s will be more accurate.

File: crates/core/tests/test_checkpoint.rs [hunk 7]
```diff
@@ -186,3 +186,387 @@ fn test_mtp_sparse_chain() {
...
+
+/// Test `insert` should panic if trying to replace genesis with a different block.
+#[test]
+#[should_panic(expected = "inserted data implies different genesis")]
+fn checkpoint_insert_cannot_replace_genesis() {
+    let block_0 = TestBlock {
+        blockhash: hash!("block_0"),
+        prev_blockhash: hash!("genesis_parent"),
+    };
+    let block_1 = TestBlock {
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs]: Isn't this redundant for this test case?

File: crates/core/tests/test_checkpoint.rs [hunk 8]
```diff
@@ -186,3 +186,387 @@ fn test_mtp_sparse_chain() {
...
+
+/// Test `insert` should panic if inserted data's prev_blockhash implies a different genesis.
+#[test]
+#[should_panic(expected = "inserted data implies different genesis")]
+fn checkpoint_insert_cannot_displace_genesis() {
+    let block_0 = TestBlock {
+        blockhash: hash!("block_0"),
+        prev_blockhash: hash!("genesis_parent"),
+    };
+    let block_1 = TestBlock {
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs]: Isn't this redundant for this test case?

File: crates/core/src/checkpoint.rs [hunk 9]
```diff
@@ -256,21 +284,22 @@ where
     /// Construct from an iterator of block data.
     ///
     /// Returns `Err(None)` if `blocks` doesn't yield any data. If the blocks are not in ascending
-    /// height order, then returns an `Err(..)` containing the last checkpoint that would have been
-    /// extended.
+    /// height order, or there are any `prev_blockhash` mismatches, then returns an `Err(..)`
+    /// containing the last checkpoint that would have been extended.
     pub fn from_blocks(blocks: impl IntoIterator<Item = (u32, D)>) -> Result<Self, Option<Self>> {
```
Comment (nymius) [inline crates/core/src/checkpoint.rs:295]: Doesn't this merit an `Error` and `Returns` section?

**Review by nymius — COMMENTED** [review]
Approach ACK 2c29ff1438f5c737fd87ece6e39e5ebd43b8eb49

I think the "ghost" checkpoint approach is good enough for managing the new "not-really-in-chain" checkpoints.

After detaining myself for a while trying to understand for the first time the inner workings of `merge_chain`, I think it is hard to read, and it's going to get hairy to maintain in the future.
I will come back with some ideas to address that.

Something similar happens with the `CheckPointEntry::prev` recursion. Some unit test will come handy in the future.

I liked the format of `TestLocalChain`, the tables were very useful.

File: crates/chain/src/local_chain.rs (hunk 3 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: "actual checkpoint" is a checkpoint that is not a placeholder, a.k.a. a checkpoint that actually contains data.

Actual checkpoints are persisted, placeholder checkpoints are not. If this is a placeholder checkpoint, we don't want to persist it as placeholder checkpoints are already implied by actual checkpoints (and there is no data to persist anyway).

File: crates/core/src/checkpoint.rs (hunk 4 above)
Comment (evanlinjin) [reply crates/core/src/checkpoint.rs:376]: No. The original chain may not have a genesis. We will not hit this if the original chain has a genesis block.

File: crates/core/tests/test_checkpoint.rs (hunk 5 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs]: Purged is *"transitively displaced"*.

In this test, `5'` has a `prev_blockhash` that displaces `4`. Since, the original chain connects `4` to `6`, if `4` is displaced, then `6` is purged.

File: crates/chain/src/local_chain.rs (hunk 3 above)
Comment (nymius) [reply crates/chain/src/local_chain.rs]: Ok, so all `Placeholder`s are "`Checkpoints` without data" and all "`Checkpoints` without data" are `Placeholders`, i.e., there is a one-to-one mapping between the wrapper iter and the actual structs.

File: crates/core/src/checkpoint.rs (hunk 4 above)
Comment (nymius) [reply crates/core/src/checkpoint.rs:376]: > The original chain may not have a genesis.

There is some no-buggy reason this could happen?

Comment (evanlinjin) [reply crates/core/src/checkpoint.rs:376]: I haven't encountered it in person, but if there's no reason to error/panic, then we shouldn't do it.

`CheckPoint` is an update type, chain sources might merge chains together before forming an update (not sure).

**Comment by evanlinjin** [comment]
> After detaining myself for a while trying to understand for the first time the inner workings of merge_chain, I think it is hard to read, and it's going to get hairy to maintain in the future.
I will come back with some ideas to address that.

@nymius thanks for looking into that! For this PR, it's good enough to just understand how `merge_chains` is meant to function, and ensure our tests covers all cases and edge cases. Just treat it like a black box for now and we can look into it later.

**Comment by nymius** [comment]
> @nymius thanks for looking into that! For this PR, it's good enough to just understand how `merge_chains` is meant to function, and ensure our tests covers all cases and edge cases. Just treat it like a black box for now and we can look into it later.

Yes, not intention to turn this into a refactor. Something for another PR.

File: crates/chain/src/local_chain.rs [hunk 10]
```diff
@@ -485,6 +494,36 @@ impl<D> FromIterator<(u32, D)> for ChangeSet<D> {
...
+                f,
+                "`prev_blockhash` doesn't match block at height {} ({})",
+                expected.height, expected.hash
+            ),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ApplyBlockError {}
```
Comment (nymius) [inline crates/chain/src/local_chain.rs]: [`core::error::Error` was stabilized in 1.81](https://doc.rust-lang.org/nightly/core/error/trait.Error.html), as MSRV is 1.85 now, I think we can remove this (and all the others too)

File: crates/chain/src/local_chain.rs [hunk 11]
```diff
@@ -594,18 +633,42 @@ fn merge_chains<D>(
...
+            debug_assert!(
+                matches!(err, ApplyBlockError::MissingGenesis),
+                "PrevBlockhashMismatch should never happen"
+            );
+            CannotConnectError {
+                try_include_height: 0,
+            }
+        })?;
+        Ok((new_tip, changeset))
+    }
```
Comment (nymius) [inline crates/chain/src/local_chain.rs:663]: This is not improving the comprehension in a meaningful way. I would remove this function and add the comment directly on top of the code, once moved to the end.

File: crates/chain/tests/test_local_chain.rs [hunk 12]
```diff
@@ -953,3 +1016,296 @@ proptest! {
...
+        // chain  | _       C(prev=B)  D
+        // update | _   A              D
+        // result | _   A              D
+        // ```
+        //
+        // Both chains agree on D at height 4. Chain has C at height 2 with `prev_blockhash = B`,
+        // but no B exists. Update introduces A at height 1, which displaces C because
+        // C's `prev_blockhash` ("B") doesn't match A's hash ("A").
+        //
+        // Note: This can only happen if chains are constructed incorrectly.
```
Comment (nymius) [inline crates/chain/tests/test_local_chain.rs]: If this can only happen if chains are constructed incorrectly, shouldn't the expected behavior be to fail then?

File: crates/core/tests/test_checkpoint.rs [hunk 13]
```diff
@@ -186,3 +186,379 @@ fn test_mtp_sparse_chain() {
...
+    let result = cp.clone().push(101, conflicting_block);
+
+    assert!(
+        result.is_err(),
+        "push should fail when prev_blockhash conflicts"
+    );
+    assert!(
+        result.unwrap_err().eq_ptr(&cp),
+        "should return self on error"
+    );
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs:458]: To give a full picture of the expected behavior I would check the chain still contains block 100.

File: crates/core/tests/test_checkpoint.rs [hunk 14]
```diff
@@ -186,3 +186,379 @@ fn test_mtp_sparse_chain() {
...
+
+    // Push at height 101 with matching prev_blockhash
+    let result = cp.push(101, matching_block);
+
+    assert!(
+        result.is_ok(),
+        "push should succeed when prev_blockhash matches"
+    );
+    let new_cp = result.unwrap();
+    assert_eq!(new_cp.height(), 101);
```
Comment (nymius) [inline crates/core/tests/test_checkpoint.rs:490]: Check that `new_cp.len()` is 2.

**Review by nymius — COMMENTED** [review]
The code is addressing the referenced issues, and the changes correlate with the discussion on them.
I haven't found any conceptual concerns, but would like to hear other voices @ValuedMammal.

P.S.: I wasn't expecting to do that many comments, next time I'll push them in a single review. Sorry

**Comment by ValuedMammal** [comment]
I'd prefer if we moved the checkpoint entry feature to a separate PR and just focus this one on validating prev_blockhash connections.

**Comment by evanlinjin** [comment]
> I'd prefer if we moved the checkpoint entry feature to a separate PR and just focus this one on validating prev_blockhash connections.

@ValuedMammal `CheckPointEntry` is actually the mechanism through which `prev_blockhash` validation works in `merge_chains` — it fills in gaps between checkpoints using `prev_blockhash`, allowing the merge logic to detect conflicts at heights where no checkpoint exists. These two features are tightly coupled, so splitting them would mean either rewriting `merge_chains` without `CheckPointEntry` or landing `CheckPointEntry` first as a dependency PR.

File: crates/chain/src/local_chain.rs (hunk 10 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs]: Good suggestion. I think it's better to implement this in a separate PR. I created #2132

File: crates/chain/tests/test_local_chain.rs (hunk 12 above)
Comment (evanlinjin) [reply crates/chain/tests/test_local_chain.rs]: If the original chain is already in a bad state, erroring would prevent recovery. The merge should be best-effort.

File: crates/core/src/checkpoint.rs (hunk 9 above)
Comment (evanlinjin) [reply crates/core/src/checkpoint.rs:295]: Added in 6191afcd6bfafd90d1eaaedc66b8c72202a5cdb2

File: crates/core/tests/test_checkpoint.rs (hunk 13 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs:458]: Done in 6191afcd6bfafd90d1eaaedc66b8c72202a5cdb2

File: crates/core/tests/test_checkpoint.rs (hunk 14 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs:490]: Added in 6191afcd6bfafd90d1eaaedc66b8c72202a5cdb2

File: crates/chain/src/local_chain.rs (hunk 11 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:663]: I'd prefer to keep `finish` — it's called from two places and contains a `debug_assert!` that documents an important invariant. Inlining would mean duplicating the assert + error mapping at both call sites. The named function also communicates intent — "we're done iterating, apply the result."

File: crates/chain/src/local_chain.rs (hunk 2 above)
Comment (evanlinjin) [reply crates/chain/src/local_chain.rs:501]: This is already addressed — `ConflictingGenesis` was removed. The current enum only has `MissingGenesis` and `PrevBlockhashMismatch`, each with a single semantic.

File: crates/core/tests/test_checkpoint.rs (hunk 6 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs]: Addressed — both checkpoints are now explicitly checked (height + hash) in addition to the count.

File: crates/core/tests/test_checkpoint.rs (hunk 7 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs]: Addressed — the redundant blocks have been removed from these tests.

File: crates/core/tests/test_checkpoint.rs (hunk 8 above)
Comment (evanlinjin) [reply crates/core/tests/test_checkpoint.rs]: Addressed — same as above.

File: crates/chain/src/local_chain.rs (hunk 11 above)
Comment (nymius) [reply crates/chain/src/local_chain.rs:663]: Good point, leave as it is.

File: crates/chain/tests/test_local_chain.rs (hunk 12 above)
Comment (nymius) [reply crates/chain/tests/test_local_chain.rs]: So merge has the chances of fixing the original chain state?

**Review by nymius — APPROVED** [review]
ACK 3a101d801d493da89ec0897ccf610a423d3504d5

Needs rebase.

File: crates/chain/tests/test_local_chain.rs (hunk 12 above)
Comment (evanlinjin) [reply crates/chain/tests/test_local_chain.rs]: Yes. The alternative is worse.

Failing loudly on a state that is, by assumption, reachable but malformed means: 
* (a) you need a separate repair API for users to dig themselves out,
* (b) the enum grows a variant nobody wants to handle,
* (c) callers have to match on it to do the thing the library could have just done for them.

**Comment by evanlinjin** [comment]
self-ACK c486ba75865806bbe1d88c646385ac57df611ef5

@nymius Thanks for the review. I've rebased, got Claude to review the codebase one more time and fixed/improved docs and a few small fixes.
